cola/widgets/branch.py

```diff
@@ -0,0 +1,520 @@
+
+            if self.current_branch is not None and tracked_branch is not None:
+                status = {'ahead': 0, 'behind': 0}
+                status_str = ""
+                args = ["--oneline"]
+
+                origin = tracked_branch + ".." + self.current_branch
+                log = self.git_helper.log(origin, args)
+                status['ahead'] = len(log[1].splitlines())
+
+                origin = self.current_branch + ".." + tracked_branch
+                log = self.git_helper.log(origin, args)
+                status['behind'] = len(log[1].splitlines())
+
+                if status["ahead"] > 0:
+                    status_str += "\u2191" + str(status["ahead"])
+
+                if status["behind"] > 0:
+                    status_str += "  \u2193" + str(status["behind"])
+
+                if status_str != "":
+                    item.setText(0, item.text(0) + "\t" + status_str)
+
+    def git_action_async(self, action, args, kwarg=None, refresh_tree=False,
+                         update_remotes=False):
+        if kwarg is None:
+            kwarg = {}
+
+        task = AsyncGitActionTask(self, self.git_helper, action, args, kwarg,
+                                  refresh_tree, update_remotes)
+        progress = standard.ProgressDialog(N_('Executing action %s') % action,
+                                           N_('Updating'), self)
+
+        self.runtask.start(task, progress=progress,
+                           finish=self.git_action_completed)
+
+    def git_action_completed(self, task):
+        status, out, err = task.result
+        self.git_helper.show_result(task.action, status, out, err)
+
+        if task.refresh_tree:
+            self.refresh()
+
+        if task.update_remotes:
+            model = main.model()
+            model.update_remotes()
+
+    def fetch_action(self):
+        branch = self.tree_helper.get_full_name(self.selected_item(),
+                                                SEPARATOR_CHAR)
+        remote_branch = gitcmds.tracked_branch(branch)
+
+        if remote_branch is not None:
+            rgx = re.compile(r'^(?P<remote>[^/]+)/(?P<branch>.+)$')
+            match = rgx.match(remote_branch)
+
+            if match:
+                remote = match.group('remote')
+                branch_name = match.group('branch')
+                self.git_action_async('fetch', [remote, branch_name])
+
+    def push_action(self):
+        branch = self.tree_helper.get_full_name(self.selected_item(),
+                                                SEPARATOR_CHAR)
+        remote_branch = gitcmds.tracked_branch(branch)
+
+        if remote_branch is not None:
+            rgx = re.compile(r'^(?P<remote>[^/]+)/(?P<branch>.+)$')
+            match = rgx.match(remote_branch)
+
+            if match:
+                remote = match.group('remote')
+                branch_name = match.group('branch')
+                self.git_action_async('push', [remote, branch_name])
```

Comment:
Just want to make sure I'm following -- the assumption in the "Push" action here is that, if the selected branch is `origin/master` and we select "Push" then we want to push the local "master" branch to the remote.  That probably warrants an in-code comment.